### PR TITLE
Remove single-view assumption from ScrollPhysics

### LINKED ISSUE #117478: Remove hard-coded single-view assumption from `ScrollPhysics`
`ScrollPhysics` currently has hard-coded single view assumptions in two places, see the `WidgetsBinding.instance.window` in the snippet below:

https://github.com/flutter/flutter/blob/d1244b7c9b8d676a84bdba2a440ddae1e0907a8c/packages/flutter/lib/src/widgets/scroll_physics.dart#L386-L391

https://github.com/flutter/flutter/blob/d1244b7c9b8d676a84bdba2a440ddae1e0907a8c/packages/flutter/lib/src/widgets/scroll_physics.dart#L249-L258

In a multi-viewed world, this will not work anymore because the `ScrollPhysics` have no idea what view/window the scrollable they control is in. We need to refactor this, i.e. either the ScrollPhysics somehow need to look up the current view via `View.of(context)` or the relevant information need to be passed in.

### PR DESCRIPTION
**Pull request by goderbauer**
Fixes https://github.com/flutter/flutter/issues/117478.
Part of https://github.com/flutter/flutter/issues/116929.

This PR replaces the `tolerance` getter on `ScrollPhysics` with a `tolleranceFor(ScrollMetrics metrics)` method, which allows the `ScrollPhysics` to calculate the tolerance based on the current `ScrollMetrics`. Furthermore, the `ScrollMetrics` mixin is extended to include a getter for the `devicePixelRatio`, which represents the dpr of the `FlutterView` into which the `Scrollable` associated with the `ScrollMetrics` is drawn into.

Overall, this enables `ScrollPhysics` to calculate the tolerance based on the dpr of the View the Scrollable in question is drawn into and removes the assumption from `ScrollPhysics` that `WidgetsBinding.instance.window` is the only window.

For backwards compatibility, the (deprecated) `ScrollPhysics.tolerance` getter is kept. It forwards to the new `tolleranceFor` method using a fake `ScrollMetrics` object.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 1]
```diff
@@ -727,6 +727,12 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
   }
 
   void _handleDragCancel() {
+    if (_gestureDetectorKey.currentContext == null) {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollable.dart:730]: Is this change related?

**Comment by Piinks** [comment]
For reference, dart fix will not be (currently) possible here: https://github.com/dart-lang/sdk/issues/50699

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/scrollable.dart:730]: Yeah, without this, the `test/widgets/scrollable_dispose_test.dart` fails because as part of cancelling the hold or drag below the scrollable wants to go ballistic, for which it needs to know the device pixel ratio for one of its calculations. Doing that now requires an inherited widget lookup, which fails if the we are in the process of being disposed. That's fine though, since we can just skip going ballistic if we get disposed anyways.

**Comment by goderbauer** [comment]
> For reference, dart fix will not be (currently) possible here: https://github.com/dart-lang/sdk/issues/50699

I think that is ok. Writing a custom ScrollPhysics seems to be rather rare, so I think we should be ok with having people migrate manually. Whenever the issue gets fixed in dartfix, we can add one, though.

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scrollable.dart:730]: Oh I see, thank you!

**Review by Piinks — APPROVED** [review]
LGTM!
